### ISSUE TITLE
E3700 false positive when Stages entries are Fn::If with AWS::NoValue

### ISSUE THREAD

**Issue by zfLQ2qx2** [aws-cloudformation/cfn-lint#4672]
### CloudFormation Lint Version

1.56.0

### What operating system are you using?

Amazon Linux 2023

### Describe the bug

E3700 treats a stage's index in the `Stages` list as its order. When a `Stages`
entry is an `Fn::If` that can resolve to `AWS::NoValue`, index and order are not
the same thing, and every conditional source stage past the first gets flagged.

The pattern is one conditional stage per source type, which is how you get
CodeCommit/GitHub/S3 support out of a single template:

```yaml
Stages:
  - !If [IsCodeCommit, <source stage>, !Ref 'AWS::NoValue']
  - !If [IsGitHub,     <source stage>, !Ref 'AWS::NoValue']
  - Name: Build
    ...
```

`_is_first_stage` skips `Fn::If` by advancing two path elements and then returns
True only if it lands on index 0, so the second entry is classified as a later
stage and its `Category: Source` is rejected.

The rule never looks at conditions. In the template below the two conditions are
mutually exclusive by construction, `SourceType` only takes one value, so exactly
one source stage is ever realized and it is always first. E3700 still fires.
E3701 stays silent on the same template, so the condition scenarios are
available where the rule runs, it just doesn't use them.

Seven templates here hit it, `PipelineType` V1 and V2 alike.

### Expected behavior

No E3700 on the template below.

Either resolve condition scenarios the way E3701 does, or have
`_is_first_stage` return `None` when the position depends on an `Fn::If` that can
be `AWS::NoValue`, since the index no longer tells you the order.

### Reproduction template

```yaml
AWSTemplateFormatVersion: '2010-09-09'
Parameters:
  SourceType:
    Type: String
    AllowedValues: [CodeCommit, GitHub]
Conditions:
  IsCodeCommit: !Equals [!Ref SourceType, CodeCommit]
  IsGitHub: !Equals [!Ref SourceType, GitHub]
Resources:
  Pipeline:
    Type: AWS::CodePipeline::Pipeline
    Properties:
      RoleArn: 'arn:aws:iam::123456789012:role/pipeline'
      ArtifactStore:
        Type: S3
        Location: artifact-bucket
      Stages:
        - !If
          - IsCodeCommit
          - Name: SourceFromCodeCommit
            Actions:
              - Name: Source
                ActionTypeId:
                  Category: Source
                  Owner: AWS
                  Provider: CodeCommit
                  Version: '1'
                Configuration:
                  RepositoryName: repo
                  BranchName: main
                OutputArtifacts:
                  - Name: SourceArtifact
          - !Ref 'AWS::NoValue'
        - !If
          - IsGitHub
          - Name: SourceFromGitHub
            Actions:
              - Name: Source
                ActionTypeId:
                  Category: Source
                  Owner: AWS
                  Provider: CodeStarSourceConnection
                  Version: '1'
                Configuration:
                  ConnectionArn: 'arn:aws:codeconnections:us-east-1:123456789012:connection/11111111-1111-1111-1111-111111111111'
                  FullRepositoryId: org/repo
                  BranchName: main
                OutputArtifacts:
                  - Name: SourceArtifact
          - !Ref 'AWS::NoValue'
        - Name: Build
          Actions:
            - Name: Build
              ActionTypeId:
                Category: Build
                Owner: AWS
                Provider: CodeBuild
                Version: '1'
              Configuration:
                ProjectName: project
              InputArtifacts:
                - Name: SourceArtifact
              OutputArtifacts:
                - Name: BuildArtifact
        - Name: Deploy
          Actions:
            - Name: Deploy
              ActionTypeId:
                Category: Deploy
                Owner: AWS
                Provider: CloudFormation
                Version: '1'
              Configuration:
                ActionMode: CREATE_UPDATE
                StackName: stack
                TemplatePath: 'BuildArtifact::template.yaml'
              InputArtifacts:
                - Name: BuildArtifact
```

```
$ cfn-lint repro.yaml
E3700 'Source' is not one of ['Build', 'Approval', 'Deploy', 'Test', 'Invoke', 'Compute']
repro.yaml:40:19
```
